### PR TITLE
feat(alerts): common field support for on-demand metrics

### PR DESCRIPTION
**Pull request by obostjancic**
Part of [Allow common fields in alert creation](https://github.com/getsentry/sentry/issues/52503)

### CODE REVIEW

**Review by iambriccardo — APPROVED** [review]

File: src/sentry/snuba/metrics/extraction.py [hunk 1]
```diff
@@ -182,11 +201,39 @@ def is_on_demand_snuba_query(snuba_query: SnubaQuery) -> bool:
...
+
+        return False
+    except InvalidSearchQuery as e:
+        logger.error(f"Failed to parse search query: {query}. Error: {e}")
+        return False
+
+
+def _is_on_demand_search_filter(token: QueryToken) -> bool:
+    if isinstance(token, SearchFilter):
+        return token.key.name not in _STANDARD_METRIC_FIELDS
```
Comment (jan-auer) [inline src/sentry/snuba/metrics/extraction.py:231]: Alert queries always have `event.type:transaction` in them, but they can still be compatible with the standard metrics. In ``, we remove this component from the query, but `is_on_demand_query` can be called outside of that. 

I'm not sure whether we should primarily focus on the dataset or the `event.type` filter to determine whether this fundamentally is a performance query. However, this filter should not trigger on-demand metrics per se. 

The simplest solution would probably be to add `event.type` to `_STANDARD_METRIC_FIELDS`.

**Review by jan-auer — CHANGES_REQUESTED** [review]

File: src/sentry/snuba/metrics/extraction.py [hunk 2]
```diff
@@ -182,11 +201,39 @@ def is_on_demand_snuba_query(snuba_query: SnubaQuery) -> bool:
...
+
+    if isinstance(token, ParenExpression):
+        return any([_is_on_demand_search_filter(c) for c in token.children])
```
Comment (iambriccardo) [inline src/sentry/snuba/metrics/extraction.py:234]: Just a nit, you could rework the recursive implementation of `_contains_on_demand_search_filters` by splitting the function in two:
- `_query_has_on_demand()` -> which parses the query and calls `_contains_on_demand_search_filters` which just loops over the tokens.
- `_contains_on_demand_search_filters` -> which loops over a list of tokens.

Then you could have this like:
```suggestion
        return _contains_on_demand_search_filters(token.children)
```

Comment (iambriccardo) [reply src/sentry/snuba/metrics/extraction.py:234]: With this you will not implement the same logic twice. What do you think?

Comment (obostjancic) [reply src/sentry/snuba/metrics/extraction.py]: Yep, good idea. Done in [b0f91a2](https://github.com/getsentry/sentry/pull/52612/commits/b0f91a261a7b266dea4982002ed38ab36c6b3ada)

File: src/sentry/snuba/metrics/extraction.py (hunk 1 above)
Comment (obostjancic) [reply src/sentry/snuba/metrics/extraction.py]: Added `event.type` to `_STANDARD_METRIC_FIELDS` for now. We can revisit it later

File: src/sentry/snuba/metrics/extraction.py [hunk 3]
```diff
@@ -182,11 +201,39 @@ def is_on_demand_snuba_query(snuba_query: SnubaQuery) -> bool:
...
     if not dataset or not query:
         return False
 
-    return Dataset(dataset) == Dataset.PerformanceMetrics and "transaction.duration" in query
+    return Dataset(dataset) == Dataset.PerformanceMetrics and _contains_on_demand_search_filters(
+        query
+    )
+
+
+def _contains_on_demand_search_filters(query: str) -> bool:
```
[the remark this answers is not in the document]
Comment (obostjancic) [reply src/sentry/snuba/metrics/extraction.py]: Refactored this logic in [b0f91a2](https://github.com/getsentry/sentry/pull/52612/commits/b0f91a261a7b266dea4982002ed38ab36c6b3ada)
Renamed the functions as suggested and moved case 3 (incompatible query) up to `is_on_demand_query`

**Review by jan-auer — APPROVED** [review]

**Pull request merged**